p2p/net/swarm/swarm_stream_test.go

```diff
@@ -0,0 +1,45 @@
+package swarm
+
+import (
+	"context"
+	"sync/atomic"
+	"testing"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/stretchr/testify/require"
+)
+
+type asyncStreamWrapper struct {
+	network.MuxedStream
+	beforeClose func()
+}
+
+func (s *asyncStreamWrapper) AsyncClose(onDone func()) error {
+	s.beforeClose()
+	err := s.Close()
+	onDone()
```

Comment:
Could we throw a `go` statement somewhere to exercise potential deadlocks and races ?
I think it would be doable by adding `sync.WaitGroup` inside `asyncStreamWrapper`.

We could also add a second `sync.Mutex` which would allow the test the behavior of listing streams while an `AsyncClose` is still in flight (does that deadlocks ? Can we still find the open stream ?).

Reply:
I updated it with test for 100 concurrent stream Close with GetStreams check. Is this what you had in mind?

Reply:
I didn't thought of having more than one stream running at once but I think it's good idea.

My point was just about calling onDone out-of-order.